docs/environment-variables.md
# Backend environment variables

For backend deployment and development environments, the following environment variables
should be set. They can be set by creating the file `backend/.env` and adding the following
contents.

> [!NOTE]
> In production environments, do not include the variables in the "DEVELOPMENT VARIABLES" section.

> [!IMPORTANT]
> The `DJANGO_SECRET_KEY` variable must be set to a random value for the backend to work and be secure.
> Execute the provided command to generate a suitable value, and put it in between the double-quotes after `DJANGO_SECRET_KEY=`.

```shell
# Generate a secret key for the backend by running this command:
#     uv run python -c "from django.core.management.utils import get_random_secret_key; print(get_random_secret_key())"
DJANGO_SECRET_KEY=""

# The base URL that the frontend will be accessible at.
FRONTEND_BASE_URL=http://localhost:3000

# These are the PostgreSQL database credentials used for the portal backend.
PORTAL_DB_DATABASE=portal
PORTAL_DB_USER=portal
PORTAL_DB_PASSWORD=portal
PORTAL_DB_HOST=localhost
PORTAL_DB_PORT=5432

# If you want GitHub login to work, set these environment variables according
# to the details for your GitHub OAuth app.
GITHUB_CLIENT_ID=
GITHUB_CLIENT_SECRET=

# If you want the backend server to send emails, set these environment variables
# to the appropriate SMTP credentials.
EMAIL_HOST=
EMAIL_HOST_USER=
EMAIL_HOST_PASSWORD=

# The reset-database script will connect to the PostgreSQL server specified by
# PORTAL_DB_HOST and PORTAL_DB_PORT and will use the following credentials to
# create the database and user specified with PORTAL_DB_DATABASE,
# PORTAL_DB_USER, and PORTAL_DB_PASSWORD.
#
# These will also be used as the default database, username, and password
# for the PostgreSQL container in the Docker Compose production deployment.
PG_CONNECT_DATABASE=postgres
PG_CONNECT_USER=postgres
PG_CONNECT_PASSWORD=postgres

##########################################################################
##########################################################################
###                                                                    ###
###  DEVELOPMENT VARIABLES                                             ###
###                                                                    ###
###  The environment variables below are only relevant in development  ###
###  environments. Remove them in production environments.             ###
###                                                                    ###
##########################################################################
##########################################################################

DJANGO_DEBUG=1

##########################################################################
##########################################################################
###                                                                    ###
###  END DEVELOPMENT VARIABLES                                         ###
###                                                                    ###
##########################################################################
##########################################################################
```

# Frontend environment variables

For frontend deployment and development environments, the following environment variables
should be set. They can be set by creating the file `frontend/.env` and adding the following
contents.

```shell
# The base URL for the backend API.
# In the development environment, this runs on a separate origin from the frontend -- locally on port 8000.
# In the recommended production environment, this can be set to /api as it is hosted on the same origin.
VITE_API_BASE_URL=http://127.0.0.1:8000/api

# The URL for the backend admin panel.
# In the development environment, this runs on a separate origin from the frontend -- locally on port 8000.
# In the recommended production environment, this can be set to /admin as it is hosted on the same origin.
VITE_ADMIN_URL=http://127.0.0.1:8000/admin

# If you want GitHub login to work, set this to the client ID of the GitHub OAuth app.
# It should be the same as the GITHUB_CLIENT_ID environment variable set on the backend.
VITE_GITHUB_CLIENT_ID=
```


backend/config/settings.py
"""
Django settings for config project.

Generated by 'django-admin startproject' using Django 3.2.7.

For more information on this file, see
https://docs.djangoproject.com/en/3.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.2/ref/settings/
"""

import os
from pathlib import Path
from urllib.parse import urlparse

from environ import Env

# This is the backend directory
# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

# reading .env file
env = Env()
env.read_env(env_file=BASE_DIR / ".env")

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

SECRET_KEY = None

secret_key_file_path = "/run/secrets/django-secret-key"
if os.path.isfile(secret_key_file_path):
    with open(secret_key_file_path, "rt") as f:
        SECRET_KEY = f.read().strip()

    if len(SECRET_KEY) == 0:
        raise Exception(f"{secret_key_file_path} is empty")

if SECRET_KEY is None:
    # Raises django's ImproperlyConfigured exception if SECRET_KEY not in os.environ
    SECRET_KEY = env("DJANGO_SECRET_KEY")

# False if not in os.environ
DEBUG = env("DJANGO_DEBUG", default=False)

FRONTEND_BASE_URL = env("FRONTEND_BASE_URL", default="http://localhost:3000")
parsed_frontend_base_url = urlparse(FRONTEND_BASE_URL)

frontend_host = parsed_frontend_base_url.hostname
ALLOWED_HOSTS = [
    frontend_host,
] + (["localhost", "127.0.0.1", "[::1]"] if DEBUG else [])


# Application definition

INSTALLED_APPS = [
    "rest_framework",
    "rest_framework.authtoken",
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "portal",
    "corsheaders",
    "drf_spectacular",
    "django_admin_listfilter_dropdown",
]

MIDDLEWARE = [
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "config.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR / "templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "config.wsgi.application"


# Database
# https://docs.djangoproject.com/en/3.2/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": env("PORTAL_DB_DATABASE", default="portal"),
        "USER": env("PORTAL_DB_USER", default="portal"),
        "PASSWORD": env("PORTAL_DB_PASSWORD", default="portal"),
        "HOST": env("PORTAL_DB_HOST", default="localhost"),
        "PORT": env("PORTAL_DB_PORT", default="5432"),
    }
}


# User substitution
# https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#auth-custom-user
AUTH_USER_MODEL = "portal.User"

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "America/Edmonton"

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

STATIC_URL = "/djangostatic/"
STATIC_ROOT = os.path.join(BASE_DIR, "djangostatic/")

MEDIA_URL = "/media/"
MEDIA_ROOT = os.path.join(BASE_DIR, "build", "media")

# Default primary key field type
# https://docs.djangoproject.com/en/3.2/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

if DEBUG:
    CORS_ALLOW_ALL_ORIGINS = True
else:
    frontend_origin = (
        f"{parsed_frontend_base_url.scheme}://{parsed_frontend_base_url.netloc}"
    )
    CORS_ALLOWED_ORIGINS = [frontend_origin]

REST_FRAMEWORK = {
    "DEFAULT_SCHEMA_CLASS": "drf_spectacular.openapi.AutoSchema",
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "rest_framework.authentication.TokenAuthentication",
    ],
    "DEFAULT_RENDERER_CLASSES": [
        "rest_framework.renderers.JSONRenderer",
    ]
    + (["rest_framework.renderers.BrowsableAPIRenderer"] if DEBUG else []),
}

SPECTACULAR_SETTINGS = {
    "TITLE": "Capstone Projects Portal API",
    "DESCRIPTION": "This is the API for the Capstone Projects Portal.",
    "VERSION": "1.0.0",
}

# Setting for email functionality
EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
EMAIL_HOST = env("EMAIL_HOST", default="")
EMAIL_PORT = 587
EMAIL_USE_TLS = True
EMAIL_HOST_USER = env("EMAIL_HOST_USER", default="")
EMAIL_HOST_PASSWORD = env("EMAIL_HOST_PASSWORD", default="")

# GitHub OAuth2 keys
GITHUB_CLIENT_ID = env("GITHUB_CLIENT_ID", default="")
GITHUB_CLIENT_SECRET = env("GITHUB_CLIENT_SECRET", default="")

# Test runner
TEST_RUNNER = "config.test_runner.TestRunner"
